src/utilities/modules/vaccines/aggregation.js

```diff
@@ -0,0 +1,212 @@
+import {
+  linkSensorLogToItemBatches,
+  createFullAggregateSensorLogs,
+  createGenericSensorLog,
+} from './utilities';
+
+/**
+ * Aggregation methods for SensorLogs. PreAggregate
+ * and FullAggregate.
+ *
+ * Preaggregates are aggregations of raw data synced
+ * from a sensor. Sensors can be quite sensitive to
+ * temperatures, so a shorter interval is used on a
+ * physical sensor to achieve a more representitive
+ * temperature once aggregated.
+ *
+ * Full aggregations are much larger consolidations
+ * of sensor logs in which a breach has not occurred,
+ * to limit the number of SensorLog records as well
+ * as SensorLogItemBatchJoins.
+ */
+
+const PREAGGREGATE_TYPE = 'preAggregate';
+const FULL_AGGREGATE_TYPE = 'aggregate';
+
+const ONE_MINUTE_MILLISECONDS = 1000 * 60;
+const ONE_HOUR_MILLISECONDS = ONE_MINUTE_MILLISECONDS * 60;
+const FIVE_MINUTES_MILLISECONDS = ONE_MINUTE_MILLISECONDS * 5;
+const EIGHT_HOURS_MILLISECONDS = ONE_HOUR_MILLISECONDS * 8;
+
+const NO_FULL_AGGREGATE_PERIOD = EIGHT_HOURS_MILLISECONDS * 4 + FIVE_MINUTES_MILLISECONDS;
+const FULL_AGGREGATION_INTERVAL = EIGHT_HOURS_MILLISECONDS + FIVE_MINUTES_MILLISECONDS;
+const PREAGGREGATE_INTERVAL = 20 * ONE_MINUTE_MILLISECONDS;
+/**
+ * Aggregates sensor logs that have been pulled from a physical sensor.
+ */
+export function preAggregateLogs({ sensor, database }) {
+  const { location } = sensor;
+  let logsToDelete = [];
+  const logsToAdd = [];
+  const sortedLogs = sensor.sensorLogs
+    .filtered('aggregation == null || aggregation == ""')
+    .sorted('timestamp');
+  let firstTimestamp = sortedLogs[0].timestamp;
+  let endTimestamp = new Date(firstTimestamp.getTime() + PREAGGREGATE_INTERVAL);
+  let logsToDeleteTemp = [];
+  let temperature = Infinity;
+  sortedLogs.forEach(sensorLog => {
+    const { timestamp: logTimestamp, temperature: logTemperature } = sensorLog;
+    // If this sensorlog does exceed the aggregates end timestamp, check if it has
+    // the minimum temperature and push it onto the potential delete stack, to be
+    // deleted if it is included in an aggregation.
+    if (logTimestamp <= endTimestamp) {
+      logsToDeleteTemp.push({ ...sensorLog });
+      if (temperature > sensorLog.temperature) temperature = logTemperature;
+    } else {
+      // If the currentlogs timestamp is greater than the end timestamp of the current
+      // aggregation, increment the timestamps for the next aggregation.
+      while (logTimestamp >= new Date(endTimestamp)) {
+        firstTimestamp = endTimestamp;
+        endTimestamp = new Date(endTimestamp.getTime() + PREAGGREGATE_INTERVAL);
+      }
+      logsToDelete = [...logsToDelete, ...logsToDeleteTemp];
+      logsToDeleteTemp = [{ ...sensorLog }];
```

Comment:
why not [sensorLog] ?

Reply:
Because realm are live records I quite like using clones 

ce984a71cc3e5878294bf06f16ce5f86ce5a820b